### PR TITLE
Local unit changes modal

### PR DESCRIPTION
**Pull request by shreeyash07**
## Addresses:
- #1495 
## Depends on: 
- https://github.com/IFRCGo/go-api/pull/2346
## Changes
- Add changes modal in the local unit form
- find changes logic between the value change and local unit response

## This PR doesn't introduce:
- [x] typos
- [x] conflict markers
- [x] unwanted comments
- [x] temporary files, auto-generated files or secret keys
- [x] `console.log` meant for debugging
- [x] codegen errors

### CODE REVIEW

File: app/src/utils/common.ts [hunk 1]
```diff
@@ -52,3 +52,28 @@ export function getFirstTruthyString(
...
+            return { ...acc, ...flattenObject(value as Record<string, unknown>, newKey) };
+        }
+        return { ...acc, [newKey]: value };
+    }, {} as Record<string, unknown>);
+}
+
+export function getLastSegment(str: string, delimiter: string) {
+    const parts = str.split(delimiter);
+    return parts[parts.length - 1];
+}
```
Comment (AdityaKhatri) [inline app/src/utils/common.ts]: write tests

File: app/src/utils/common.ts [hunk 2]
```diff
@@ -52,3 +52,28 @@ export function getFirstTruthyString(
 
     return invalidText;
 }
+
+export function compareArrays(newArray: unknown[], oldArray: unknown[]): boolean {
+    if (newArray.length !== oldArray.length) {
+        return false;
+    }
+    return newArray.every((id) => oldArray.includes(id));
+}
```
Comment (tnagorra) [inline app/src/utils/common.ts]: Either the comparison is no correct or the name of the function is not correct.

File: app/src/utils/common.ts [hunk 3]
```diff
@@ -52,3 +52,28 @@ export function getFirstTruthyString(
...
+    prefix?: string,
+): Record<string, unknown> {
+    return Object.entries(inputObject).reduce((acc, [key, value]) => {
+        const newKey = prefix ? `${prefix}.${key}` : key;
+        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
+            return { ...acc, ...flattenObject(value as Record<string, unknown>, newKey) };
+        }
+        return { ...acc, [newKey]: value };
+    }, {} as Record<string, unknown>);
+}
```
Comment (tnagorra) [inline app/src/utils/common.ts:76]: Does this not handle array elements?

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 4]
```diff
@@ -94,6 +119,53 @@ function FormGrid(props: FormGridProps) {
...
+
+    Object.keys(flattenedValues).forEach((key) => {
+        const newValue = flattenedValues[key];
+        const oldValue = flattenedOldValues[key];
+
+        if ((newValue === undefined || newValue === null)
+            && (oldValue === undefined || oldValue === null)
+        ) {
+            return;
+        }
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Let's just use isNotDefined

File: app/src/utils/common.ts [hunk 5]
```diff
@@ -52,3 +52,28 @@ export function getFirstTruthyString(
 
     return invalidText;
 }
+
+export function compareArrays(newArray: unknown[], oldArray: unknown[]): boolean {
```
Comment (AdityaKhatri) [inline app/src/utils/common.ts]:
```suggestion
export function isArraySame(newArray: unknown[], oldArray: unknown[]): boolean {
```

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 6]
```diff
@@ -94,6 +119,53 @@ function FormGrid(props: FormGridProps) {
     );
 }
 
+function getChangedFormFields(
+    newValues: PartialLocalUnits,
+    oldValues: LocalUnitResponse,
+    formFieldOptions: LocalUnitOptions,
+) {
+    const flattenedValues = flattenObject(newValues);
+    const flattenedOldValues = flattenObject(oldValues);
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:133]: I am not sure if we can use this approach. This fails if/when we have array elements.

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 7]
```diff
@@ -198,18 +297,90 @@ function LocalUnitsForm(props: Props) {
...
+                phone,
+                email,
+                city_en,
+                city_loc,
+                link,
+                location_json,
+                health: {
+                    affiliation: health?.affiliation,
+                    functionality: health?.functionality,
+                    health_facility_type: health?.health_facility_type,
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:366]: Any reason to write these out individually? Why not just spread this out?

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 8]
```diff
@@ -330,6 +502,26 @@ function LocalUnitsForm(props: Props) {
...
 
+    const onDoneButtonClick = useCallback(
+        () => {
+            if (isDefined(localUnitDetailsResponse) && isDefined(localUnitsOptions)) {
+                const changedFormFields = getChangedFormFields(
+                    value,
+                    localUnitDetailsResponse,
+                    localUnitsOptions,
+                );
+                setLocalUnitChangedFormFields(() => changedFormFields);
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]:
```suggestion
                setLocalUnitChangedFormFields(changedFormFields);
```

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 9]
```diff
@@ -94,6 +119,53 @@ function FormGrid(props: FormGridProps) {
     );
 }
 
+function getChangedFormFields(
+    newValues: PartialLocalUnits,
+    oldValues: LocalUnitResponse,
+    formFieldOptions: LocalUnitOptions,
+) {
```
Comment (AdityaKhatri) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Write return type
```suggestion
function getChangedFormFields(
    newValues: PartialLocalUnits,
    oldValues: LocalUnitResponse,
    formFieldOptions: LocalUnitOptions,
): { key: string; value: string }[] {
```

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/useLocalUnitFormFieldLabels.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
+import { useTranslation } from '@ifrc-go/ui/hooks';
+
+import i18n from './i18n.json';
+
+interface Props {
+    key: string;
+}
+function useLocalUnitFormFieldLabels(props: Props) {
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/useLocalUnitFormFieldLabels.ts:8]: Are these created just to display the labels on diff mode?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/useLocalUnitFormFieldLabels.ts (hunk 10 above)
Comment (shreeyash07) [reply app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/useLocalUnitFormFieldLabels.ts:8]: Yes.

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx (hunk 7 above)
Comment (shreeyash07) [reply app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:366]: when we compare the value later, the comparison function checks for `_display` values. We don't need these values for form fields. So we didn't spread them out.

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx (hunk 6 above)
Comment (shreeyash07) [reply app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:133]: we will refactor this later. will add a TODO

File: app/src/utils/common.ts (hunk 3 above)
Comment (shreeyash07) [reply app/src/utils/common.ts:76]: we will refactor this later. will add a TODO

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 11]
```diff
@@ -94,6 +119,53 @@ function FormGrid(props: FormGridProps) {
...
+                const valueLabel = options.find(
+                    (option: Option) => option.id === newValue,
+                )?.name;
+                changedValues.push({ key, value: valueLabel });
+            } else if (typeof newValue === 'boolean') {
+                changedValues.push({ key, value: newValue ? 'Yes' : 'No' }); // TODO: use translations
+            } else {
+                changedValues.push({ key, value: newValue as string });
+            }
+        }
```
Comment (AdityaKhatri) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: add few comments on what is being done to make it more readable

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 12]
```diff
@@ -94,6 +119,53 @@ function FormGrid(props: FormGridProps) {
     );
 }
 
+function getChangedFormFields(
```
Comment (AdityaKhatri) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:127]: Write tests

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx (hunk 8 above)
Comment (AdityaKhatri) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]:
```suggestion
                setLocalUnitChangedFormFields(changedFormFields);
```

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: app/src/utils/common.ts (hunk 1 above)
Comment (samshara) [reply app/src/utils/common.ts]: Added a `TODO` comment.

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**